### PR TITLE
Store `RawKNNRegressor` as component in `TransformedKNeighborsRegressor`

### LINKED ISSUE #100: Performance issues using Hamming distance in tree-based estimators
In #85, we implemented the `RFNNRegressor` to mimic similar functionality in `yaImpute` and in #98, we introduced the `WeightedHammingDistanceMixin` to serve as the `kneighbors` distance metric which overcame some numeric instability found in `scipy.spatial.distance.hamming` owing to the use of `np.dot` in the function (detailed in [this thread](https://github.com/lemma-osu/sknnr/pull/98#issuecomment-3207781962)).  

When experimenting with using `RFNNRegressor` in a spatial context using `sklearn-raster`, we noticed very slow runtimes even over small spatial domains (e.g. 50 x 50 windows) and relatively small number of forests and trees used to create the `RFNNRegressor` estimator.  Through profiling (with both `cProfile` and `line_profiler`), the bottleneck appears to be the element-wise calculation of Hamming distance through `kneighbors`.  For example, a 50 x 50 window with 2000 samples across 100 estimators (trees) in a single forest has to do (2500 * 2000 ~= 4.3 million) python calls through `sklearn`.

Ideally, fewer Python calls for distance finding would be better.  For a single query point `q`, we can create the node dissimilarity matrix with the reference node IDs `R` with a single numpy call (i.e. `diff = R != q`).  This code sample illustrates the current implementation (using a single `RandomForestClassifier`) versus a much faster vectorized approach.
```
# In ipython

from sklearn.datasets import make_classification
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsRegressor
from scipy.spatial.distance import hamming

X, y = make_classification(
    n_samples=2000,
    n_features=10,
    n_informative=6,
    n_redundant=0,
    random_state=0,
    shuffle=False
)
clf = RandomForestClassifier(n_estimators=100, max_depth=10, random_state=0)
clf = clf.fit(X, y)

# Create the reference node matrix (R)
R = clf.apply(X)

# Get a sample query point (q)
q = clf.apply(X[0].reshape(1, -1))

# Current method using a kNN regressor
# Note that I'm using `scipy.spatial.distance.hamming` as a proxy for our
# custom `WeightedHammingDistanceMetric` as precision isn't critical for
# this test
knn_est = KNeighborsRegressor(n_neighbors=5, metric="hamming").fit(R, y)

# Get timing on getting the top 5 neighbors/distances
# 1.72 ms ± 85.6 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
%timeit knn_est.kneighbors(q)

# New method comparing q against all R
# 9.13 ns ± 0.624 ns per loop (mean ± std. dev. of 7 runs, 100,000,000 loops each)
w = np.ones(R.shape[1], dtype=np.int32)
%timeit "diff = R != q; d = diff @ w; np.argpartition(d, 5)[:5]"

# Assert that the same neighbors are identified
old = knn_est.kneighbors(q)[:5][1]

diff = R != q
d = diff @ w
new = np.argpartition(d, 5)[:5]
print(np.all(old == new))
# True
```
This represents a massive (200,000x) speedup if I'm not doing something stupid.  

@aazuspan, I haven't totally thought this through, but what I think this means is that `RFNNRegressor` (and other tree-based estimators) at the very least need to override `TransformedKNeighborsRegressor.kneighbors` and I'm not sure there would be value continuing to have `TransformedKNNRegressor` as a base class.  Other than this small example, I haven't yet implemented this, but I don't foresee issues handling individual tree weights (already demonstrated above) or forest weights.  If you have other thoughts on this, I'd be grateful to hear them.

PS. This was the source of what I was describing last week and why you wouldn't have seen the same slowdowns just using RF and GB directly (i.e. without using `.apply()`)

### PR DESCRIPTION
**Pull request by grovduck**
This PR changes the inheritance structure of `TransformedKNeighborsRegressor` classes to include a `RawKNNRegressor` instance as an estimator attribute rather than inheriting this class.  This allows derived classes to alter the parameters of `KNeighborsRegressor` initialization at run-time after the transformer has been fit, which is necessary for Hamming distance weighting using tree-based NN methods.

Additionally, `WeightedTreesNNRegressor` using "hamming" as the distance metric and no longer uses `WeightedHammingDistanceMetric`, which reverts #98 and would close #100.

@aazuspan, definitely treat this as draft right now, but here is a first cut of using composition instead of inheritance.  It wasn't too messy, but mostly adds (rather than removes) code to get properties and call methods from the `RawKNNRegressor` instance (which is now an estimator attribute called `regressor_`) which used to be handled by inheritance.  All tests are passing locally for me, although I'm guessing we'll have regression problems from the [`np.dot`](https://github.com/lemma-osu/sknnr/issues/100#issuecomment-3663096737) issue.  I've been pretty liberal with comments throughout that we can remove later.

I think the natural place to "split" the inheritance was at `TransformedKNeighborsRegressor` given that `RawKNNRegressor` still "is-a" `KNeighborsRegressor` with just some mixins on top.  We could have done this split at `WeightedTreesRegressor` instead since I am assuming that only tree/forest based estimators will need to modify the regression parameters, so let me know what you think about that.  One thing that I don't love is that we need to reimplement a number of methods (e.g. `predict`, `score`, etc.) to redirect the calls to `self.regressor_` and right now there is some duplication associated with transforming features before calling the regressor.

**EDIT**: The regression tests are all passing which I wasn't expecting for the RFNN estimator.  I read back through our discussion in #98 and it seems like they should still be failing, but I'm loathe to give too much thought to it until it fails again.  Conversely, we're getting doc build failures that I can replicate locally and could use your help understanding.  It doesn't seem related to changes here.

### CODE REVIEW

File: src/sknnr/_base.py [hunk 1]
```diff
@@ -123,14 +148,45 @@ def _set_fitted_transformer(self, X, y):
...
+            "algorithm": self.algorithm,
+            "leaf_size": self.leaf_size,
+            "p": self.p,
+            "metric": self.metric,
+            "metric_params": self.metric_params,
+            "n_jobs": self.n_jobs,
+        }
+        reg_init_kwargs.update(self._get_additional_regressor_init_kwargs())
+        self.regressor_ = RawKNNRegressor(**reg_init_kwargs)
+        self.regressor_._set_dataframe_index_in(X)
```
[the remark this answers is not in the document]
Comment (grovduck) [reply src/sknnr/_base.py]: When `_set_dataframe_index_in` is called with `X`, `X` may be a `pd.DataFrame` with an `index` attribute.  However, `X_transformed` is guaranteed to be a `np.array` (without an `index` attribute) and so will not override `self.dataframe_index_in_`.  This is similar logic that we used when using inheritance rather than composition.

[the remark this answers is not in the document]
Comment (aazuspan) [reply src/sknnr/_base.py]: Ah, I was confused by this as well, but your explanation makes sense. Is it worth adding a comment to explain why it needs to be called explicitly with the untransformed X? Also, I don't think this makes a practical difference, but maybe it would be clearer to call `regressor_._set_dataframe_index_in` *after* `regressor_.fit` to make the precedence more obvious.

File: src/sknnr/_base.py [hunk 2]
```diff
@@ -123,14 +148,45 @@ def _set_fitted_transformer(self, X, y):
...
+            "n_neighbors": self.n_neighbors,
+            "weights": self.weights,
+            "algorithm": self.algorithm,
+            "leaf_size": self.leaf_size,
+            "p": self.p,
+            "metric": self.metric,
+            "metric_params": self.metric_params,
+            "n_jobs": self.n_jobs,
+        }
+        reg_init_kwargs.update(self._get_additional_regressor_init_kwargs())
```
[the remark this answers is not in the document]
Comment (grovduck) [reply src/sknnr/_base.py]: > Just to be clear, are you considering taking `metric_params` out of the `TransformedKNeighborsRegressor` signature? Currently, you theoretically _could_ use a custom metric and pass `metric_params` in with something like `GNNRegressor`, since it doesn't override `__init__`, right? Are you thinking that's something we should prevent? Not that I'm arguing against that, just trying to make sure I understand the suggestion and motivation.

Yes, that's where my mind was going but, in retrospect, that seems a bit heavy handed.  And you're correct that in the case of `RFNNRegressor`, `metric_params` isn't exposed to the user so this issue shouldn't come up.  So, as you say, those constraints should probably happen at the derived classes' `__init__` and leave all options available in `TransformedKNeighborsRegressor` should other yet-to-be-created estimators need them.  Just because I haven't imagined some crazy distance metric doesn't mean it won't come up in practice.

For this particular suggestion, though, I'm OK leaving as is or creating a generic dictionary updater if you can imagine that would be used at some point.

> Yeah, constraining unsupported options makes sense. I imagine that means overriding some `__init__` methods to include additional validation and/or refine their signatures?

I think that's right.  The tree-based methods are the ones where distance finding is the least flexible, so locking those down a bit more (as in `RFNNRegressor`) makes sense to me.  I think I'd like to create a separate issue for that, though.

[the remark this answers is not in the document]
Comment (aazuspan) [reply src/sknnr/_base.py]: > For this particular suggestion, though, I'm OK leaving as is or creating a generic dictionary updater if you can imagine that would be used at some point.

I'm happy with leaving it as-is.

>  I think I'd like to create a separate issue for that, though.

Sounds good!

**Comment by aazuspan** [comment]
Getting the docstrings more fleshed out is great! I haven't had a chance to review yet, but had a few responses on mkdocs configuration.

> One could hide these using mkdocstrings.handlers.python.options.filters, but as far as I know, this would also impact methods which we don't want to hide.

Yeah, I see what you mean that these aren't adding much, but I agree that manually filtering them sounds tricky. I'm surprised there's not a configuration option available, but I didn't turn anything up. It looks like you *can* remove that redundancy by documenting the attributes directly instead of including them in the class docstring.

For example, I tried changing the `MahalanobisTransformer` docstring to:

```python
class MahalanobisTransformer(OneToOneFeatureMixin, TransformerMixin, BaseEstimator):
    """
    Transformer that first standardizes features to zero mean and unit variance
    using a modified `StandardScaler` that uses N-1 degrees of freedom, then
    applies a Mahalanobis transformation to decorrelate features.

    When used in conjunction with Euclidean distance metrics, this is
    equivalent to calculating
    [Mahalanobis distances](https://en.wikipedia.org/wiki/Mahalanobis_distance)
    between samples.
    """

    n_features_in_ : int
    """The number of features seen during `fit`."""
    scaler_ : StandardScalerWithDOF
    """The fitted scaler used to standardize feature data."""
    transform_ : np.ndarray
    """The Mahalanobis transformation matrix attribute."""
```

and it now only renders the attributes once:

image

That affects the docstrings you see in the IDE too, i.e. attributes aren't shown when you hover over the class anymore, but now they *are* shown when you hover over an attribute, and you also get type information.

I'm not sure if we want to go that route or not. I can live with the redundancy if you don't want to dive into that decision now.

>  LIkewise, if a class relies on inheritance from a different package (e.g. RawKNNRegressor.predict, it doesn't show up in the docstrings.

After some poking around, I was able to include inherited methods using [`preload_modules`](https://mkdocstrings.github.io/python/usage/configuration/general/#preload_modules):

```diff
# mkdocs.yml

            merge_init_into_class: true
            signature_crossrefs: true
+            preload_modules:
+            - sklearn
```

I also added `scikit-learn` to the dependencies for `tool.hatch.envs.docs` in the `pyproject.toml`, but in hindsight I think that *probably* shouldn't be necessary if it's inherited from the base environment.
